[FIRRTL] InnerSymAttr: Add per field inner symbol

prithayan opened the pull request:
This PR updates the `InnerSymAttr` to add per field inner symbols.
 `InnerSymAttr` defines the properties of the `inner_sym` attribute. It specifies the symbol
    name and symbol visibility for each field ID.
 For any ground types, there are no subfields and the field ID is 0. For aggregate types, a
    unique field ID is assigned to each field by visiting them in a depth-first pre-order.
 The custom assembly format ensures that there is no change for the default case.
If a symbol for any field other than `fieldID=0` is specified, then the format is like this:
`%w = firrtl.wire sym [<@w,2,public>,<@x,1,private>,<@syh,4,public>] : !firrtl.bundle<a: uint<1>, b: uint<1>, c: uint<1>, d: uint<1>>`
This also adds the capability to specify the visibility of each symbol.
Some followup work pending:
1. Implement verifier for the `InnerSymbol` trait.
2. Update the `InnerSymbolTable` to handle per field symbols.
3. Update all renaming logic for different passes to consider the symbols per field.

dtzWill commented:
This looks really great, I like syntax (and plan). LowerTypes will want to scatter these when expanding, right? Just curious, if putting a symbol on an interior or non-leaf node of a bundle tree (of fields), how will that be lowered, where do the symbol(s) go?

Looking it over, few comments (apologies for not adding as review/attaching to source): There are some commented-out bits that probably should be dropped.

Not sure about handling of the case when we have multiple symbols/props and one with fieldID=0, can you add a test for that? It'd be good to at least check the intended notation works for that, please. In particular, when there's a symbol for fieldID=0, it looks like this prints something like:

@foo[<@w,2,public>]



(good so far!)

But I don't think the parser will handle that presently ("else if" -> "else"?). It will also allow parsing it when put in the array ( <@foo,0,public> ) as an alternative form, but I'm not sure there's benefit being strict there since it's clear what is meant? (Be liberal in what you accept, strict in what you produce?)

On a related note, visibility on the fieldID=0 case seems assumed to be public (?) unless written within the square bracket notation. And in that case, I don't think non-public visibility on that will round-trip presently?

dtzSiFive commented:
This looks really great, I like syntax (and plan). LowerTypes will want to scatter these when expanding, right? Just curious, if putting a symbol on an interior or non-leaf node of a bundle tree (of fields), how will that be lowered, where do the symbol(s) go?

Looking it over, few comments (apologies for not adding as review/attaching to source): There are some commented-out bits that probably should be dropped.

Not sure about handling of the case when we have multiple symbols/props and one with fieldID=0, can you add a test for that? It'd be good to at least check the intended notation works for that, please. In particular, when there's a symbol for fieldID=0, it looks like this prints something like:

@foo[<@w,2,public>]



(good so far!)

But I don't think the parser will handle that presently ("else if" -> "else"?). It will also allow parsing it when put in the array ( <@foo,0,public> ) as an alternative form, but I'm not sure there's benefit being strict there since it's clear what is meant? (Be liberal in what you accept, strict in what you produce?)

On a related note, visibility on the fieldID=0 case seems assumed to be public (?) unless written within the square bracket notation. And in that case, I don't think non-public visibility on that will round-trip presently? 

EDIT: oh I see you mention if there are more fields they all should be square bracket notation. Okay, then modulo ensuring we print it that way, I think there's still a detail re: visibility of fieldID=0 to work out, I guess its how to print/parse when single symbol on root/ground that's not public, which presently we'd have to create ourselves but does seem like something we want to support?

prithayan commented:
> This looks really great, I like syntax (and plan). LowerTypes will want to scatter these when expanding, right? Just curious, if putting a symbol on an interior or non-leaf node of a bundle tree (of fields), how will that be lowered, where do the symbol(s) go?
> 

I'm considering, maybe we never add a symbol to the `fieldID=0` or the entire bundle. And always add symbols to the required field or all the fields. Currently there is no use case for this, so we can define the semantics and later extend it if required.
So, if there is a `DontTouch` on the bundle, we can add symbols to each field instead of a symbol to the `fieldID=0`.
To answer your question, LowerTypes now is much simpler, because there is a 1-1 relation between a symbol and its user. We never have to blow out a symbol and then update its users. So essentially LowerTypes does not have to do anything.

> Looking it over, few comments (apologies for not adding as review/attaching to source): There are some commented-out bits that probably should be dropped.
> 

I think I fixed them.

> Not sure about handling of the case when we have multiple symbols/props and one with fieldID=0, can you add a test for that? It'd be good to at least check the intended notation works for that, please. In particular, when there's a symbol for fieldID=0, it looks like this prints something like:
> 
> @foo[<@w,2,public>]
> 
> (good so far!)
> 

Yes, I fixed that, and added a testcase.
> But I don't think the parser will handle that presently ("else if" -> "else"?). It will also allow parsing it when put in the array ( <@foo,0,public> ) as an alternative form, but I'm not sure there's benefit being strict there since it's clear what is meant? (Be liberal in what you accept, strict in what you produce?)
> 

So, by default if there is a single symbol and its on `fieldID=0` and its public, then it is printed as `sym @name`. 
Anything else is printed in the standard format, `[<fieldID,name,visibility>,...]`. We can iterate over what's the best format to print them.

> On a related note, visibility on the fieldID=0 case seems assumed to be public (?) unless written within the square bracket notation. And in that case, I don't think non-public visibility on that will round-trip presently?
> 

Fixed this.
> EDIT: oh I see you mention if there are more fields they all should be square bracket notation. Okay, then modulo ensuring we print it that way, I think there's still a detail re: visibility of fieldID=0 to work out, I guess its how to print/parse when single symbol on root/ground that's not public, which presently we'd have to create ourselves but does seem like something we want to support?

Added a test case for this: 
` %w2 = firrtl.wire sym [<@w2,0,private>]`
